selfie.c

```diff
@@ -4053,12 +4198,741 @@ int main_emulator(int argc, int *argv, int *cstar_argv) {
+    }
+
+    process = list_entry_get_data(list_pop_back(g_readyqueue));
+
+    if(DEBUG_1) {
+        printString('/','/','/','/','/','/','/','/','/','/','/','/','/','/','/','/','/','/','/',10);
+        printString('n','e','x','t',' ','p','r','o','c','e','s','s',' ','t','o',' ','r','u','n',' ');
+        printInt(process_get_id(process));
+        putchar(10);
+    }
+
+    return process;
+}
+
+// Switches to PROCESS
+void process_switch(int *process) {
+
+    if(process == g_running_process)
+        return;
+
+    // Save the state of the current process
+    process_set_pc(g_running_process, pc);
+    process_set_registers(g_running_process, registers);
+    process_set_memory(g_running_process, memory);
+    process_set_reg_hi(g_running_process, reg_hi);
+    process_set_reg_lo(g_running_process, reg_lo);
+
+    list_push_front(g_readyqueue, g_running_process);
+
+    // Switch to PROCESS
+    pc = process_get_pc(process);
+    registers = process_get_registers(process);
+    memory = process_get_memory(process);
+    reg_hi = process_get_reg_hi(process);
+    reg_lo = process_get_reg_lo(process);
+
+    // We get a new running process
+    g_running_process = process;
+}
+
+// NUM ... Number of bytes
+int *memcpy(int *source, int num) {
+    int i;
+    int *dest;
+
+    i = 0;
+    dest = (int*)malloc(num * 4);
+    while(i < num) {
+        *(dest + i) = *(source + i);
+        i = i + 1;
+    }
+
+    return dest;
+}
+
+void printInt(int i) {
+    print(itoa(i, string_buffer, 10, 0));
+}
+
+int int_length(int i) {
+    int j;
+    int length;
+
+    if(i == 0)
+        return 1;
+    
+    if(i < 0)
+        i = (-1)*i;
+
+    j = 1;
+    length = 0;
+    while(i > j - 1) {
+        j = j * 10;
+        length = length + 1;
+    }
+
+    return length;
+}
+
+// -----------------------------------------------------------------
+// ------------------------------ LIST -----------------------------
+// -----------------------------------------------------------------
+
+int *list_init() {
+    int *list;
+    int size;
+    int *head;
+    int *tail;
+    // list:
+    // +----+-------------+
+    // |  0 | ptr to head |
+    // |  1 | ptr to tail |
+    // |  2 | size        |
+    // +----+-------------+
+
+    list = (int*)malloc(3 * 4);
+    list_set_head(list, 0);
+    list_set_tail(list, 0);
+    list_set_size(list, 0);
+
+    return list;
+}
+
+// Getters
+int *list_get_head(int *list) {
+    return (int*)*list;
+}
+
+int *list_get_tail(int *list) {
+    return (int*)*(list + 1);
+}
+
+int list_get_size(int *list) {
+    return *(list + 2);
+}
+
+// Setters
+void list_set_head(int *list, int *head) {
+    *list = (int)head;
+}
+
+void list_set_tail(int *list, int *tail) {
+    *(list + 1) = (int)tail;
+}
+
+void list_set_size(int *list, int size) {
+    *(list + 2) = size;
+}
+
+// Inserts DATA at HEAD
+void list_push_front(int *list, int *data) {
+    int *newEntry;
+    int *head;
+    int size;
+
+    // list entry:
```

Comment:
the data layout of your list nodes is good. Placing the next and previous pointer at the beginning of the node allows for a generalisation of the list: Your list could handle all kinds of objects that follow your convention, i.e., first word is a next pointer, second word is a prev pointer, third word and beyond is the data. Think of that advantage! Instead of passing a pointer to the data in the argument list, you could pass a pointer to any node allocated outside of this function.